### ISSUE TITLE
 java.lang.NoSuchMethodException: org.apache.ibatis.reflection.MetaObject.forObject(java.lang.Object)

### ISSUE THREAD

**Issue by ghostman-git** [pagehelper/Mybatis-PageHelper#277]
Caused by: java.lang.NoSuchMethodException: org.apache.ibatis.reflection.MetaObject.forObject(java.lang.Object)
	at java.lang.Class.getDeclaredMethod(Class.java:2130) ~[?:1.8.0_181]
	at com.github.pagehelper.util.MetaObjectUtil.<clinit>(MetaObjectUtil.java:47) ~[pagehelper-5.1.5.jar:?]
	at com.github.pagehelper.util.PageObjectUtil.getPageFromObject(PageObjectUtil.java:83) ~[pagehelper-5.1.5.jar:?]
	at com.github.pagehelper.page.PageParams.getPage(PageParams.java:79) ~[pagehelper-5.1.5.jar:?]
	at com.github.pagehelper.PageHelper.skip(PageHelper.java:58) ~[pagehelper-5.1.5.jar:?]
	at com.github.pagehelper.PageInterceptor.intercept(PageInterceptor.java:102) ~[pagehelper-5.1.5.jar:?]
	at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:60) ~[mybatis-3.2.8.jar:3.2.8]
	at com.sun.proxy.$Proxy71.query(Unknown Source) ~[?:?]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:108) ~[mybatis-3.2.8.jar:3.2.8]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:102) ~[mybatis-3.2.8.jar:3.2.8]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectOne(DefaultSqlSession.java:66) ~[mybatis-3.2.8.jar:3.2.8]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_181]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_181]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_181]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_181]
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:433) ~[mybatis-spring-1.3.2.jar:1.3.2]
	at com.sun.proxy.$Proxy68.selectOne(Unknown Source) ~[?:?]
	at org.mybatis.spring.SqlSessionTemplate.selectOne(SqlSessionTemplate.java:166) ~[mybatis-spring-1.3.2.jar:1.3.2]
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:68) ~[mybatis-3.2.8.jar:3.2.8]
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:52) ~[mybatis-3.2.8.jar:3.2.8]
	at com.sun.proxy.$Proxy69.findBusinessValueByParamPort(Unknown Source) ~[?:?]
	at com.framework.ureport.fileprovider.HxatFileReportProvider.setApplicationContext(HxatFileReportProvider.java:110) ~[classes/:?]
	at org.springframework.context.support.ApplicationContextAwareProcessor.invokeAwareInterfaces(ApplicationContextAwareProcessor.java:121) ~[spring-context-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.context.support.ApplicationContextAwareProcessor.postProcessBeforeInitialization(ApplicationContextAwareProcessor.java:97) ~[spring-context-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1622) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:519) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeansOfType(DefaultListableBeanFactory.java:508) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.getBeansOfType(AbstractApplicationContext.java:1188) ~[spring-context-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at com.bstek.ureport.export.ReportRender.setApplicationContext(ReportRender.java:142) ~[ureport2-core-2.2.9.jar:?]
	at org.springframework.context.support.ApplicationContextAwareProcessor.invokeAwareInterfaces(ApplicationContextAwareProcessor.java:121) ~[spring-context-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.context.support.ApplicationContextAwareProcessor.postProcessBeforeInitialization(ApplicationContextAwareProcessor.java:97) ~[spring-context-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1622) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:312) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:308) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:351) ~[spring-beans-4.3.16.RELEASE.jar:4.3.16.RELEASE]

**Comment by miaogr86** [comment]
mybatis 什么版本？是不是版本太低？

**Comment by ghostman-git** [comment]
mybatis-3.2.8       低版本不再兼容了？

**Comment by miaogr86** [comment]
mybatis-3.2.8  不兼容

**Comment by ghostman-git** [comment]
是mybatis的版本问题， 但是对于老项目而言，升级mybatis版本是困难的，

**Comment by abel533** [comment]
@ghostman-git 分页插件支持 3.1.0+ 版本，你用的什么版本？

**Comment by ghostman-git** [comment]
@abel533   v2.2.8   没升级最新pagehelper是可以的。

**Comment by abel533** [comment]
你用的 2.2.8 版本的 mybatis?

**Comment by tosmq008** [comment]
我的  mybatis 3.4.6  也出现这个问题了，最新的 5.1.6

**Comment by abel533** [comment]
```java
public class MetaObjectUtil {
    public static Method method;

    static {
        try {
            Class<?> metaClass = Class.forName("com.github.pagehelper.util.MetaObjectWithReflectCache");
            method = metaClass.getDeclaredMethod("forObject", Object.class);
        } catch (Throwable e1) {
            try {
                //在下面这行断点看看上面抛出的 e1 异常时什么内容，能不能把异常发出来看看
                Class<?> metaClass = Class.forName("org.apache.ibatis.reflection.MetaObject");
                method = metaClass.getDeclaredMethod("forObject", Object.class);
            } catch (Exception e2) {
                throw new PageException(e2);
            }
        }

    }
```

注意上述代码注释，断点看看异常。

**Comment by yang-mr** [comment]
public static MetaObject forObject(Object object, ObjectFactory objectFactory, ObjectWrapperFactory objectWrapperFactory) {

你反射不到吧？。。。
		<mybatis.version>3.2.8</mybatis.version>

**Comment by yang-mr** [comment]
java.lang.NoSuchMethodException: org.apache.ibatis.reflection.MetaObject.forObject(java.lang.Object)
	at java.lang.Class.getDeclaredMethod(Class.java:2130)
	at com.github.pagehelper.util.MetaObjectUtil.<clinit>(MetaObjectUtil.java:47)
	at com.github.pagehelper.dialect.AbstractHelperDialect.processParameterObject(AbstractHelperDialect.java:118)
	at com.github.pagehelper.PageHelper.processParameterObject(PageHelper.java:88)
	at com.github.pagehelper.util.ExecutorUtil.pageQuery(ExecutorUtil.java:166)
	at com.github.pagehelper.PageInterceptor.intercept(PageInterceptor.java:104)
	at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:60)
	at com.sun.proxy.$Proxy54.query(Unknown Source)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:108)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:102)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

**Comment by abel533** [comment]
@yang-mr 是用的最新版本吗？最新版本测试过3.1.0+到3.4.6。

**Comment by yang-mr** [comment]
<pagehelper.version>5.1.6</pagehelper.version>
是的最新版本啊  我不是贴了源码么

**Comment by orangejz** [comment]
在mybatis3.2.5版本 中 同样发现这个问题，在这个版本中MetaObject类，只有
![image](https://user-images.githubusercontent.com/38680953/49718577-7cdd3a80-fc95-11e8-8981-1e2480567b95.png)三个参数的调用，ReflectorFactory这个工厂在这个版本中还不存在。

**Comment by zhangxinyuan1** [comment]
Could not initialize class com.github.pagehelper.util.MetaObjectUtil ？
mybatis 版本是3.2.4
pagehelper 版本是 5.1.8

**Comment by zhangxinyuan1** [comment]
@abel533 上面问题是什么原因导致的？要如何解决？

**Comment by zhangxinyuan1** [comment]
单独拿插件源码来看是看不出来的，因为没有和spring集成  最终的原因就是mybatis 和 spring-mybatis 版本太低造成的。

**Comment by abel533** [comment]
解决 MetaObject 版本兼容性问题 fixed #349
